File: Week10/P2579.py

# 계단의 개수 
n = int(input())
#계단에 적혀 있는 점수목록 
stair = [0]*n
for i in range(n):
    stair[i] = int(input())


if n == 1:
    print(stair[0])
elif n == 2:
    print(stair[0] + stair[1])
else: 

    #dp 배열 초기화
    dp = [0]*n

    #첫 계단의 최대 값은 당연이 첫 계단에 쓰여있는 점수 값 
    dp[0] = stair[0]
    # 두번째 계단의 최대 값은 바로 직전 값 더하기 
    dp[1] = stair[1] + stair[0]
    #세번째부터 생각해봐바
    dp[2] = max(dp[0] + stair[2], stair[1] + stair[2])

    for i in range (3, n):
        dp[i] = max(dp[i-3] + stair[i-1] + stair[i], dp[i-2] + stair[i])

    result = dp[-1]

    print(result)

File: Week9/P2579.java

import java.util.Scanner;



public class P2579{

    public static int get_max_score(int n, int stair[]){
        
        if(n == 1){
            return stair[0];
        }
        else if (n == 2){
            return stair[0] + stair[1];
        }
        else{
            int [] dp = new int[n];

            dp[0] = stair[0];

            dp[1] = stair[0] + stair[1];

            dp[2] = Math.max(stair[2]+stair[1] , stair[2] + stair[0]);

            for (int i = 3; i < n; i++){
                dp[i] = Math.max(dp[i-3] + stair[i-1]+ stair[i], dp[i-2] + stair[i]);

            }

            return dp[n-1];

        }
    }


    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        int [] stair = new int[n];

        for (int i = 0; i < n; i++){
            stair[i] = sc.nextInt();
        }
        int result = get_max_score(n, stair);

        System.out.println(result);
    }
}